clustermesh: fix helm etcd tls error by disabling gRPC gateway

EM-Savard opened the pull request:
When using the clustermesh.apiserver.tls.auto.method=cronJob configuration, Helm uses the cilium/certgen image to create the required clustermesh mTLS certificates. However, the generated clustermesh-apiserver-server-cert lacks the TLS Web Client Authentication extended key usage extension, which is required for proper etcd gRPC Gateway mTLS authentication.

Since Cilium uses the etcd gRPC client directly and does not require the gRPC Gateway, this fix disables the etcd gRPC Gateway on boot to avoid the certificate validation error.

The below issue explains how we were able to reproduce the following error logs in the clustermesh `etcd` container:

```
{"level":"warn","ts":"2025-12-03T17:44:33.697480Z","caller":"embed/config_logging.go:194","msg":"rejected connection on client endpoint","remote-addr":"127.0.0.1:34890","server-name":"","error":"tls: failed to verify certificate: x509: certificate specifies an incompatible key usage"}
2025/12/03 17:44:33 WARNING: [core[] [Channel #1 SubChannel #2]grpc: addrConn.createTransport failed to connect to {Addr: "0.0.0.0:2379", ServerName: "0.0.0.0:2379", }. Err: connection error: desc = "error reading server preface: remote error: tls: bad certificate"
```

Fixes: #43099

```release-note
Fix helm clustermesh etcd tls error by disabling etcd gRPC gateway
```

Feel free to comment on any missing details or code changes, it's my first time opening an issue in this project. 😄

giorio94 commented:
> What would be the lift involved in disabling the gRPC Gateway on `etcd`? Would this be seamless for all other services like _apiserver_, _kvstoremesh_, and _cilium agents_ that interact with `etcd`?

My understanding is that we never use the gRPC gateway, because all components interact with etcd via the Go client. Additionally, it is most likely not even working currently (a) because of the TLS cert issue mentioned here, and (b) because we use TLS Common Name authentication, which means that based on the following from the docs _"gRPC-gateway does not support authentication using TLS Common Name"_, we couldn't even authenticate (users don't have an associated password).

> A possible solution could be to merge this quick fix and create a separate issue/feature request to improve the etcd setup by removing the gRPC Gateway from the loop.

I'd be personally in favor of giving a try at disabling the gRPC gateway, and we can check in CI if anything unexpected emerges (I definitely don't expect that though), in which case we could revert to this patch. I don't think we are in a rush either, as the warning is not pretty, but it doesn't affect any functionality as far as I'm aware.

EM-Savard committed:
clustermesh: fix helm etcd tls error by disabling gRPC gateway

When using the clustermesh.apiserver.tls.auto.method=cronJob configuration,
Helm uses the cilium/certgen image to create the required clustermesh mTLS
certificates. However, the generated clustermesh-apiserver-server-cert lacks
the TLS Web Client Authentication extended key usage extension, which is
required for proper etcd gRPC Gateway mTLS authentication.

Since Cilium uses the etcd gRPC client directly and does not require the
gRPC Gateway, this fix disables the etcd gRPC Gateway on boot to avoid the
certificate validation error.

Fixes: #43099

Signed-off-by: Emile Savard <emile.savard@bell.ca>

MrFreezeex commented:
It is most likely not related but it seems that you would need to rebase to make the CI happy :sweat_smile: (github worfklow definitions are taken from the main branch but the actual Cilium code is from your branch which sometimes are not compatible unfortunately)

MrFreezeex commented:
/test

giorio94 commented:
> I also tried adding that parameter to the etcd-init container, but the etcdinit command doesn't support that parameter

Ah, by bad, the arguments are actually hard-coded [here](https://github.com/cilium/cilium/blob/5fd2eb19f8ab2b8350967dbc4ee9307d848c77f8/clustermesh-apiserver/etcdinit/root.go#L151-L157). Not that it matters though, given that it is short lived (it only takes care of setting up the users/roles), and because we don't enable mTLS there.
